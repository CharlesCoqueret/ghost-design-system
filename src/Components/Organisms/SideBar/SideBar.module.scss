.container {
  position: fixed;
  border-right: rgb(var(--theme-provider-chalk)) 1px solid;
  overflow: hidden auto;
  box-sizing: border-box;
  clip-path: inset(0px);

  ul {
    display: flex;
    flex-wrap: wrap;
    list-style: none;
    margin-bottom: 0;
    padding-left: 0;
    min-height: 100%;
    flex-direction: column;
    margin-block: 0px;
    padding-inline: 0px;
  }

  .sidebar {
    display: flex;
    flex-direction: column;
    margin: 0px;
    position: absolute;
    box-sizing: border-box;
    min-height: 100%;
    border-right: rgb(var(--theme-provider-chalk)) 1px solid;
    top: 0;
    left: 0;
    visibility: visible;
    transition: left 0.5s ease-in-out, visibility 0.5s linear;
  }

  .sidebarHidden {
    visibility: hidden;
    transition: left 0.5s ease-in-out, visibility 0.5s linear 0.5s;
  }

  &.unfixed {
    position: relative;
    height: stretch;
  }
}
