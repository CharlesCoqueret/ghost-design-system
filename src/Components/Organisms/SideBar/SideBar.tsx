import React, { CSSProperties, PropsWithChildren, useState } from 'react';
import classnames from 'classnames';

import { SideBarContext } from './SideBarContext';

import styles from './SideBar.module.scss';

export interface ISideBarProps {
  /** Back to menu label */
  backToMenu: string;
  /** Class for the component (optional, default: undefined) */
  className?: string;
  /** Height of the sidebar (optional, default: '100%') */
  height?: CSSProperties['height'];
  /** override css property (optional, default: undefined) */
  style?: CSSProperties;
  /** whether the sidebar can be used unfixed for demo mostyly (optional, default: false) */
  unfixed?: boolean;
  /** Width of the sidebar (optional, default: '270px')*/
  width?: CSSProperties['width'];
}

const SideBar = (props: PropsWithChildren<ISideBarProps>): React.ReactElement => {
  const { backToMenu, children, className, height, style, unfixed, width } = props;

  const [isInSubMenu, setIsInSubMenu] = useState(false);

  return (
    <SideBarContext.Provider
      value={{
        backToMenu: backToMenu,
        height: height,
        isInSubMenu: isInSubMenu,
        setIsInSubMenu: setIsInSubMenu,
        unfixed: unfixed,
        width: width,
      }}>
      <nav
        className={classnames(styles.container, className, { [styles.unfixed]: unfixed })}
        style={{ height: height, width: width, ...style }}>
        <ul
          className={classnames(styles.sidebar, { [styles.sidebarHidden]: isInSubMenu })}
          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
          style={{ width: width, left: isInSubMenu ? `-${width}` : '0px' }}>
          {children}
        </ul>
      </nav>
    </SideBarContext.Provider>
  );
};

SideBar.defaultProps = {
  className: undefined,
  height: '100%',
  style: undefined,
  unfixed: false,
  width: '270px',
};

export default SideBar;
