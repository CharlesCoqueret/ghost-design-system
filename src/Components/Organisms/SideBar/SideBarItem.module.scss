.submenu {
  display: flex;
  flex-wrap: wrap;
  list-style: none;
  margin-bottom: 0;
  padding-left: 0;
  min-height: 100%;
  flex-direction: column;
  margin-block: 0px;
  padding-inline: 0px;
}

.item {
  box-sizing: border-box;
  cursor: pointer;
  display: list-item;
  min-height: 32px;
  padding: 0px 10px;

  .label {
    display: inline-flex;
    padding: 0px 10px 0px 10px;
  }

  a {
    padding: 10px 0px;
    border-radius: 4px;
    text-decoration: none;
    display: flex;
    color: unset;

    .externalLink {
      margin: auto 10px;
      // height: 0.75rem; // fix for external link icon
    }

    .chevronRight {
      margin: auto 10px auto auto;
    }
  }

  a:hover {
    background-color: rgba(var(--theme-provider-chalk), 0.33);
  }

  a.active {
    background-color: rgba(var(--theme-provider-primary), 0.15);
    color: rgb(var(--theme-provider-primary));
  }

  &.disabled {
    color: rgb(var(--theme-provider-silver));
    pointer-events: none;
  }
}

.back {
  box-sizing: border-box;
  height: 32px;
  margin: 10px 10px 0px 10px;
  cursor: pointer;
  display: flex;
  align-items: center;
  border-radius: 4px;

  &:hover {
    background-color: rgba(var(--theme-provider-chalk), 0.33);
  }

  .label {
    display: flex;

    .chevronLeft {
      margin: auto 10px;
    }
  }
}

.submenu {
  position: fixed;
  box-sizing: border-box;
  visibility: hidden;
  transition: left 0.5s ease-in-out, visibility 0.5s linear;

  &.unfixed {
    position: absolute;
  }

  &.visible {
    visibility: visible;
    transition: left 0.5s ease-in-out, visibility 0.5s linear;
  }
}
