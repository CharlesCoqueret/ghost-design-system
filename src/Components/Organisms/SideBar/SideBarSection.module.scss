.section {
  font-weight: 600;
  min-height: 32px;
  padding-left: 10px;
  padding-right: 10px;
  padding-top: 10px;
  color: rgb(var(--theme-provider-charcoal));
  display: flex;
  align-items: center;
  box-sizing: border-box;
}

.divider {
  margin: 15px;
  border-top: rgb(var(--theme-provider-chalk)) 1px solid;
}
