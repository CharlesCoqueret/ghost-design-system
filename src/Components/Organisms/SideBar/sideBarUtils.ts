import { Location } from 'react-router-dom';

/**
 * Converts a Location of string to full path
 *
 * @param location string or Location object
 * @returns full path
 */
export const locationString = (location: string | Location): string => {
  if (typeof location === 'string') {
    return location;
  }
  return location.pathname;
};

/**
 * Indicates if the url is an external link
 *
 * @param url string or Location object
 * @returns true if the link is an external link
 */
export const isExternalLink = (url: string | Location): boolean => {
  const tmp = document.createElement('a');
  tmp.href = locationString(url);
  return tmp.host !== window.location.host;
};
