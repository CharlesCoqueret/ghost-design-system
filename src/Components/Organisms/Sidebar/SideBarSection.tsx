import React, { MouseEvent, ReactElement, useContext, useEffect, useState } from 'react';
import classnames from 'classnames';
import { NavLink, useNavigate, Location } from 'react-router-dom';

import { Icon } from '../../Atoms';
import { SideBarContext } from './SideBarContext';

export interface ISideBarItem {
  /** For test purpose only */
  dataTestId?: string;
  /** Disable the item (optional, default: false) */
  disabled?: boolean;
  /** Force external link, when internal link need to be open in another tab (optional, default: false) */
  externalLink?: boolean;
  /** Hide the item (optional, default: false) */
  hidden?: boolean;
  /** Label of the item */
  label: string;
  /** SubItems under the item (optional, default: undefined) */
  subItems?: Array<ISideBarItem>;
  /** Redicrection applied on click on item */
  to: string | Location;
}

export interface ISideBarSection {
  /** Disable the section, visible, but no interaction (optional, default: false) */
  disabled?: boolean;
  /** Label of the section (optional, default: undefined) */
  title?: string;
  /** Hide the section and the its items (optional, default: false) */
  hidden?: boolean;
  /** Add divider to the section (optional, default: false) */
  divider?: boolean;
  /** Items inside the section */
  items: Array<ISideBarItem>;
}

/**
 * Converts a Location of string to full path
 *
 * @param location string or Location object
 * @returns full path
 */
const locationString = (location: string | Location): string => {
  if (typeof location === 'string') {
    return location;
  }
  return location.pathname;
};

/**
 * Inidicates if the url is an external link
 *
 * @param url string or Location object
 * @returns true if the link is an external link
 */
const isExternalLink = (url: string | Location): boolean => {
  const tmp = document.createElement('a');
  tmp.href = locationString(url);
  return tmp.host !== window.location.host;
};

export interface ISideBarSectionProps {
  /** Items inside the section */
  section: ISideBarSection;
}

export const SideBarSection = (props: ISideBarSectionProps): ReactElement => {
  const { section } = props;
  const { divider, title, hidden, items } = section;

  return (
    <>
      {!hidden && divider && <div className='sidebar-divider' />}
      {!hidden && title && <div className='title'>{title}</div>}
      {!hidden &&
        items &&
        items.map((item) => {
          return <SideBarItem key={item.label} item={item} />;
        })}
    </>
  );
};

/**
 * Interface of navigation item component
 */
export interface ISideBarItemProps {
  item: ISideBarItem;
}

export const SideBarItem = (props: ISideBarItemProps): ReactElement => {
  const { item } = props;
  const { dataTestId, disabled, externalLink, hidden, label, subItems: subitems, to } = item;
  const { isInSubMenu, setIsInSubMenu, backToMenu } = useContext(SideBarContext);
  const [subMenuActive, setSubMenuActive] = useState(false);
  const navigate = useNavigate();

  const targetType = isExternalLink(item.to) || externalLink ? '_blank' : undefined;

  const toggleSubMenu = (): void => {
    if (!disabled) {
      setIsInSubMenu(!isInSubMenu);
      setSubMenuActive(!subMenuActive);
    }
  };

  const hasSubitems = subitems && subitems.filter((item) => !item.hidden).length > 0;
  const onClickHandler = !disabled && hasSubitems ? toggleSubMenu : undefined;

  const handleClick = (event: MouseEvent<HTMLAnchorElement>): void => {
    if (hasSubitems && !disabled) {
      event.preventDefault();
      navigate(locationString(subitems[0].to));
    }
  };

  useEffect(
    () => (): void => {
      setIsInSubMenu(false);
    },
    [],
  );

  // Is hidden
  if (hidden) return <></>;

  // Has unique subitem
  if (subitems && subitems.filter((item) => !item.hidden).length === 1) {
    return <SideBarItem item={{ label: label, to: subitems[0].to, hidden: hidden, disabled: disabled }} />;
  }

  // Is not hidden or has subitems
  return (
    <>
      <li className={`item ${disabled ? 'disabled' : ''}`} onClick={onClickHandler}>
        <NavLink
          className={classnames({ disabled: disabled })}
          data-testid={dataTestId}
          end={!subitems}
          onClick={handleClick}
          target={targetType}
          to={to}>
          <div className='label'>{label}</div>
          {externalLink && <Icon icon={['fal', 'external-link']} className='external-link' />}
          {!disabled && hasSubitems && <Icon icon={['fal', 'chevron-right']} className='chevron-right' />}
        </NavLink>
      </li>
      {hasSubitems && (
        <ul className={classnames('submenu', { 'in-submenu': subMenuActive })}>
          <li className='back' onClick={toggleSubMenu}>
            <div className='label'>
              <Icon icon={['fal', 'chevron-left']} className='chevron-left' />
              {backToMenu}
            </div>
          </li>
          <div className='sidebar-divider' />
          {subitems.map((item, index) => (
            <SideBarItem key={`item-${item.label}-${index}`} item={item} />
          ))}
        </ul>
      )}
    </>
  );
};

export default SideBarSection;
