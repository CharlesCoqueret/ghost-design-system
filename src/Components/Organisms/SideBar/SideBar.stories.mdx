import { Canvas, Meta, ArgsTable, Story, Source } from '@storybook/addon-docs';
import { MemoryRouter as Router } from 'react-router-dom';

import SideBar from './SideBar';
import SideBarItem from './SideBarItem';
import SideBarSection from './SideBarSection';

import sidebarStyle from '../../../assets/_sidebar.scss?raw';

export const Template = (args) => {
  return (
    <Router>
      <SideBar {...args}>
        <SideBarSection label='Section #1'>
          <SideBarItem label='Direct link #1' to='/link1' />
          <SideBarItem label='Direct link #2 - external link' to='/link2' externalLink />
          <SideBarItem label='Direct link #3' to='/link3' />
        </SideBarSection>
        <SideBarSection label='Section #2'>
          <SideBarItem label='Disabled subsection #2-1' to='/link4' disabled>
            <SideBarItem label='Direct link 1 in #2-1' to='/link4/link1' />
            <SideBarItem label='Direct link 2 in #2-2' to='/link4/link2' />
          </SideBarItem>
          <SideBarItem label='Subsection #2-2 with divider' to='/link5'>
            <SideBarItem label='Direct link 1 in #2-2' to='/link5/link1' />
            <SideBarItem label='Direct link 2 in #2-2' to='/link5/link2' />
          </SideBarItem>
          <SideBarItem label='Subsection #2-3' to='/link6'>
            <SideBarItem label='Direct link 1 in #2-3' to='/link6/link1' />
            <SideBarItem label='Direct link 2 in #2-3' to='/link6/link2' />
          </SideBarItem>
          <SideBarItem label='Hidden subsection #2-4' to='/link7' hidden>
            <SideBarItem label='Direct link 1 in #2-4' to='/link7/link1' />
            <SideBarItem label='Direct link 2 in #2-4' to='/link7/link2' />
          </SideBarItem>
        </SideBarSection>
        <SideBarSection divider label='Section #3 with divider'>
          <SideBarItem label='Subsection #3-1 with only one non-hidden subitem' to='/link8'>
            <SideBarItem label='Hidden direct link 1 in #3-1' to='/link8/link1' hidden />
            <SideBarItem label='Direct link 3 in #3-1' to='/link8/link2' />
          </SideBarItem>
          <SideBarItem label='Subsection #3-2 with all subitems hidden or disabled' to='/link9'>
            <SideBarItem label='Hidden direct link 1 in #3-2' to='/link9/link1' hidden />
            <SideBarItem label='Hidden direct link 2 in #3-2' to='/link9/link2' disabled />
          </SideBarItem>
        </SideBarSection>
        <SideBarSection divider>
          <SideBarItem label='Direct link in invisible section with divider' to='/link10' />
        </SideBarSection>
      </SideBar>
    </Router>
  );
};

<Meta
  title='Organism'
  component={SideBar}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
/>

# SideBar

SideBar provides access to a second level of navigation. When available, it must always be on-screen.

It is recommended to be used:

- When a page has a lot of content
- When a page which contains sub-pages
- Quick navigation between unrelated destinations

Play with SideBar live on [CodeSandbox](https://codesandbox.io/s/sidebar-demo-t4s312)

## SideBar with section

<Canvas isExpanded={true}>
  <Router>
    <SideBar backToMenu='Back to menu' height='300px' unfixed width='250px'>
      <SideBarSection label='Section #1'>
        <SideBarItem label='Direct link #1' to='/link1' />
        <SideBarItem label='Direct link #2 - external link' to='/link2' externalLink />
        <SideBarItem label='Direct link #3' to='/link3' />
      </SideBarSection>
    </SideBar>
  </Router>
</Canvas>

## SideBar with nested items

<Canvas isExpanded={true}>
  <Router>
    <SideBar backToMenu='Back to menu' height='250px' unfixed width='400px'>
      <SideBarItem label='With subitems' to='/link1'>
        <SideBarItem label='Direct link #1' to='/link1/link1' />
        <SideBarItem label='Direct link #2' to='/link1/link2' />
      </SideBarItem>
      <SideBarItem label='Direct Link' to='/link2' />
    </SideBar>
  </Router>
</Canvas>

## SideBar advanced example

<Canvas isExpanded={true}>
  <Router>
    <SideBar backToMenu='Back to menu' height='600px' unfixed width='250px'>
      <SideBarSection label='Section #1'>
        <SideBarItem label='Direct link #1' to='/link1' />
        <SideBarItem label='Direct link #2 - external link' to='/link2' externalLink />
        <SideBarItem label='Direct link #3' to='/link3' />
      </SideBarSection>
      <SideBarSection label='Section #2'>
        <SideBarItem label='Disabled subsection #2-1' to='/link4' disabled>
          <SideBarItem label='Direct link 1 in #2-1' to='/link4/link1' />
          <SideBarItem label='Direct link 2 in #2-2' to='/link4/link2' />
        </SideBarItem>
        <SideBarItem label='Subsection #2-2 with divider' to='/link5'>
          <SideBarItem label='Direct link 1 in #2-2' to='/link5/link1' />
          <SideBarItem label='Direct link 2 in #2-2' to='/link5/link2' />
        </SideBarItem>
        <SideBarItem label='Subsection #2-3' to='/link6'>
          <SideBarItem label='Direct link 1 in #2-3' to='/link6/link1' />
          <SideBarItem label='Direct link 2 in #2-3' to='/link6/link2' />
        </SideBarItem>
        <SideBarItem label='Hidden subsection #2-4' to='/link7' hidden>
          <SideBarItem label='Direct link 1 in #2-4' to='/link7/link1' />
          <SideBarItem label='Direct link 2 in #2-4' to='/link7/link2' />
        </SideBarItem>
      </SideBarSection>
      <SideBarSection label='Section #3 with divider' divider>
        <SideBarItem label='Subsection #3-1 with only one non-hidden subitem' to='/link8'>
          <SideBarItem label='Hidden direct link 1 in #3-1' to='/link8/link1' hidden />
          <SideBarItem label='Direct link 3 in #3-1' to='/link8/link2' />
        </SideBarItem>
        <SideBarItem label='Subsection #3-2 with all subitems hidden or disabled' to='/link9'>
          <SideBarItem label='Hidden direct link 1 in #3-2' to='/link9/link1' hidden />
          <SideBarItem label='Hidden direct link 2 in #3-2' to='/link9/link2' disabled />
        </SideBarItem>
      </SideBarSection>
      <SideBarSection divider>
        <SideBarItem label='Direct link in invisible section with divider' to='/link10' />
      </SideBarSection>
    </SideBar>
  </Router>
</Canvas>

## Props

<ArgsTable of={SideBar} sort='requiredFirst' />

For layout management, please refer to this [example](https://codesandbox.io/s/size-management-sidebar-xtpkk0)

## Story

<Canvas>
  <Story
    name='SideBar'
    args={{
      backToMenu: 'Back to menu',
      height: '600px',
      unfixed: true,
      width: '300px',
    }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Style

[source](https://github.com/CharlesCoqueret/ghost-design-system/blob/main/src/assets/_sidebar.scss)

<Source language='css' dark format={true} code={sidebarStyle} />
