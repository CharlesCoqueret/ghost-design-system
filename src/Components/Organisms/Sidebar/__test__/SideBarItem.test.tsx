import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SideBarItem } from '../SideBarSection';
import SideBar from '../SideBar';

describe('SideBarItem Component', () => {
  it('SideBarItem renders', () => {
    const { container } = render(
      <SideBar backToMenu={'Back to menu'} style={{ height: '600px' }}>
        <SideBarItem
          item={{
            label: 'LABEL',
            to: '/link',
            subItems: [
              {
                label: 'HIDDEN',
                to: '/link',
                hidden: true,
              },
              {
                label: 'EXTERNAL',
                to: '/link',
                externalLink: true,
              },
              {
                label: 'DISABLED',
                to: '/link',
                disabled: true,
              },
              {
                label: 'NORMAL',
                to: '/link',
              },
            ],
          }}
        />
      </SideBar>,
    );

    expect(container).toMatchSnapshot();
  });

  it('SideBarItem renders with only one subitem visible', () => {
    const { container } = render(
      <SideBar backToMenu={'Back to menu'} style={{ height: '600px' }}>
        <SideBarItem
          item={{
            label: 'LABEL',
            to: '/link',
            subItems: [
              {
                label: 'HIDDEN',
                to: '/link',
                hidden: true,
              },
              {
                label: 'LINK',
                to: '/link',
              },
            ],
          }}
        />
      </SideBar>,
    );

    expect(container).toMatchSnapshot();
  });

  it('SideBarItem handles click on entry in main menu and handles click on submenu', async () => {
    console.info = jest.fn();

    const { container } = render(
      <SideBar backToMenu={'Back to menu'} style={{ height: '600px' }}>
        <SideBarItem
          item={{
            label: 'LABEL',
            to: '/link',
            dataTestId: 'PARENT-TEST-ID',
            subItems: [
              {
                label: 'LINK1',
                dataTestId: 'CHILD-TEST-ID',
                to: '/link/link1',
              },
              {
                label: 'LINK2',
                to: '/link/link2',
              },
            ],
          }}
        />
      </SideBar>,
    );

    expect(container).toMatchSnapshot();

    const parentEntry = await screen.findByTestId('PARENT-TEST-ID');

    userEvent.click(parentEntry);
    expect(container).toMatchSnapshot();

    const childEntry = await screen.findByTestId('CHILD-TEST-ID');
    userEvent.click(childEntry);
    expect(container).toMatchSnapshot();
    expect(console.info).toHaveBeenCalledTimes(1);
    expect(console.info).toHaveBeenCalledWith('url pushed:', '/link/link1');
  });

  it('SideBarItem renders with url as Location', () => {
    const { container } = render(
      <SideBar backToMenu={'Back to menu'} style={{ height: '600px' }}>
        <SideBarItem
          item={{
            label: 'LABEL',
            to: '/link',
            subItems: [
              {
                label: 'LINK',
                to: {
                  pathname: 'PATHNAME',
                  search: 'SEARCH',
                  hash: 'HASHVALUE',
                  state: [],
                  key: 'KEY',
                },
              },
              {
                label: 'Link2',
                to: '/link2',
              },
            ],
          }}
        />
      </SideBar>,
    );

    expect(container).toMatchSnapshot();
  });

  it('SideBarItem should generate an error when not in a SideBar component', () => {
    console.error = jest.fn();

    const { unmount, container } = render(
      <SideBarItem
        item={{
          label: 'LABEL',
          to: '/link',
          subItems: [
            {
              label: 'HIDDEN',
              to: '/link',
              hidden: true,
            },
            {
              label: 'LINK',
              to: '/link',
            },
          ],
        }}
      />,
    );

    expect(container).toMatchSnapshot();

    unmount();

    expect(console.error).toBeCalledWith('Sidebar component should wrap that component');

    expect(container).toMatchSnapshot();
  });
});
