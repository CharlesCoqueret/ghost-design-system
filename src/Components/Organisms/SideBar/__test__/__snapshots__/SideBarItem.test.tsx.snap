// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBarItem Component SideBarItem handles click on entry in main menu and handles click on submenu 1`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 270px;"
  >
    <ul
      class="sidebar"
      style="width: 270px; left: 0px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          data-testid="PARENT-TEST-ID"
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu"
        style="width: 270px; left: 270px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            data-testid="CHILD-TEST-ID"
            href="/link/link1"
          >
            full url
            <div
              class="label"
            >
              LINK1
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link/link2"
          >
            full url
            <div
              class="label"
            >
              LINK2
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem handles click on entry in main menu and handles click on submenu 2`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 270px;"
  >
    <ul
      class="sidebar sidebarHidden"
      style="width: 270px; left: -270px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          data-testid="PARENT-TEST-ID"
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu visible"
        style="width: 270px; left: 0px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            data-testid="CHILD-TEST-ID"
            href="/link/link1"
          >
            full url
            <div
              class="label"
            >
              LINK1
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link/link2"
          >
            full url
            <div
              class="label"
            >
              LINK2
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem handles click on entry in main menu and handles click on submenu 3`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 270px;"
  >
    <ul
      class="sidebar sidebarHidden"
      style="width: 270px; left: -270px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          data-testid="PARENT-TEST-ID"
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu visible"
        style="width: 270px; left: 0px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            data-testid="CHILD-TEST-ID"
            href="/link/link1"
          >
            full url
            <div
              class="label"
            >
              LINK1
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link/link2"
          >
            full url
            <div
              class="label"
            >
              LINK2
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem renders 1`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 400px;"
  >
    <ul
      class="sidebar"
      style="width: 400px; left: 0px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu"
        style="width: 400px; left: 400px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            href="/link"
            target="_blank"
          >
            full url
            <div
              class="label"
            >
              EXTERNAL
            </div>
            <i
              class="icon externalLink"
              icon="fal,external-link"
            />
          </a>
        </li>
        <li
          class="item disabled"
        >
          <a
            class="disabled"
            href="/link"
          >
            full url
            <div
              class="label"
            >
              DISABLED
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link"
          >
            full url
            <div
              class="label"
            >
              NORMAL
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem renders unfixed 1`] = `
<div>
  <nav
    class="container unfixed"
    style="height: 600px; width: 400px;"
  >
    <ul
      class="sidebar"
      style="width: 400px; left: 0px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu unfixed"
        style="width: 400px; left: 400px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            href="/link"
            target="_blank"
          >
            full url
            <div
              class="label"
            >
              EXTERNAL
            </div>
            <i
              class="icon externalLink"
              icon="fal,external-link"
            />
          </a>
        </li>
        <li
          class="item disabled"
        >
          <a
            class="disabled"
            href="/link"
          >
            full url
            <div
              class="label"
            >
              DISABLED
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link"
          >
            full url
            <div
              class="label"
            >
              NORMAL
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem renders with only one subitem visible 1`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 270px;"
  >
    <ul
      class="sidebar"
      style="width: 270px; left: 0px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem renders with url as Location 1`] = `
<div>
  <nav
    class="container"
    style="height: 600px; width: 270px;"
  >
    <ul
      class="sidebar"
      style="width: 270px; left: 0px;"
    >
      <li
        class="item"
      >
        <a
          class=""
          href="/link"
        >
          partial url
          <div
            class="label"
          >
            LABEL
          </div>
          <i
            class="icon chevronRight"
            icon="fal,chevron-right"
          />
        </a>
      </li>
      <ul
        class="submenu"
        style="width: 270px; left: 270px;"
      >
        <li
          class="back"
        >
          <div
            class="label"
          >
            <i
              class="icon chevronLeft"
              icon="fal,chevron-left"
            />
            Back to menu
          </div>
        </li>
        <div
          class="divider"
        />
        <li
          class="item"
        >
          <a
            class=""
            href="PATHNAME"
          >
            full url
            <div
              class="label"
            >
              LINK
            </div>
          </a>
        </li>
        <li
          class="item"
        >
          <a
            class=""
            href="/link2"
          >
            full url
            <div
              class="label"
            >
              Link2
            </div>
          </a>
        </li>
      </ul>
    </ul>
  </nav>
</div>
`;

exports[`SideBarItem Component SideBarItem should generate an error when not in a SideBar component 1`] = `
<div>
  <li
    class="item"
  >
    <a
      class=""
      href="/link"
    >
      partial url
      <div
        class="label"
      >
        LABEL
      </div>
    </a>
  </li>
</div>
`;

exports[`SideBarItem Component SideBarItem should generate an error when not in a SideBar component 2`] = `<div />`;
